### ISSUE TITLE
`in_jupyter`

### ISSUE THREAD

**Issue by prisae** [banesullivan/scooby#8]
scooby has the `scooby.in_jupyter` and `scooby.in_ipython`. Unfortunately it is not that simple. Particularly, as the kernels are not only used by the notebooks, but also by other frontends, e.g., QtConsole:

![Jupyter QtConsole_001](https://user-images.githubusercontent.com/8020943/60160579-f2ee5c80-97f5-11e9-904d-b25d44f7527e.png)

I tried many things in the beginning of the `Versions`, but all the googling did not help. I do not know any way how to define if you are actually in a notebook or not. Or, in other words, if it can render html or not. So the QtConsole uses the kernel, but doesn't render html, for instance.

Long story short: The `scooby.in_jupyter`-promise is wrong. What do you think?

**Comment by banesullivan** [comment]
@prisae - could you do me a favor: start a Jupyter Qt console and print `globals()` and share that here

**Comment by prisae** [comment]
![Jupyter QtConsole_002](https://user-images.githubusercontent.com/8020943/60161221-5c229f80-97f7-11e9-8150-aa1ddeebc07f.png)

**Comment by prisae** [comment]
Also affects `report.PythonInfo.python_environment`

**Comment by banesullivan** [comment]
@prisae: could you try this function in QTconsole?

```py
def in_jupyter():
    class foo:
        def _repr_html_(self):
            return "html"

        def __repr__(self):
            return "plain"
    try:
        from IPython.core.interactiveshell import InteractiveShell
        format = InteractiveShell.instance().display_formatter.format
        format_dict, md_dict = format(foo(),)
        if 'text/html' in list(format_dict.keys()):
            return True
    except:
        pass
    return False
```

**Comment by prisae** [comment]
Still returns `True`,  unfortunately.

**Comment by banesullivan** [comment]
And what happens if you output an object with HTML formatting in QT console? does it render it?

can you screenshot:

```py
import scooby
scooby.investigate()
```

**Comment by prisae** [comment]
```
format_dict
Out[8]: {'text/plain': 'plain', 'text/html': 'html'}
```
The thing is, it always contains both, and only when displaying then it uses the appropriate version...

**Comment by banesullivan** [comment]
> The thing is, it always contains both, and only when displaying then it uses the appropriate version...

Weird, it didn't have `text/html` in my IPyton

**Comment by prisae** [comment]
Not in IPython, but when you have a kernel (Notebook, QtConsole, etc)

**Comment by prisae** [comment]
(I am pretty sure you have qtconsole installed too. Just type `jupyter qtconsole` instead of `jupyter notebook`.)

**Comment by prisae** [comment]
I googled the internet up and down back then, and I could not find a way to define if you are in a notebook or something else that uses the kernel. Maybe there has been something new, or you are luckier with you search, but it is likely that there is no way to decide. So renaming might be the best.

You can still document that having a kernel means probably in 99% of the cases that you are in a notebook. It is just not guaranteed...

**Comment by prisae** [comment]
We had already some discussions back when I first included the script into SimPEG: https://github.com/simpeg/simpeg/pull/698

**Comment by banesullivan** [comment]
> You can still document that having a kernel means probably in 99% of the cases that you are in a notebook. It is just not guaranteed...

That's what I'm leaning towards at this point... 

However, there *has* to be a way to do this... I'll do some hacking into the IPython and Jupyter source to see where the decision to use HTM output formatting vs. plain text is used. The code for where that decision is being made will likely give us some clues on how to implement a solution

**Comment by prisae** [comment]
Maybe. The problem is that, as far as I got, this decision is made client-side (so not in the kernel, but by the notebook or by qtconsole); as scooby is a module there is probably no way to test client-side behaviour; it is agnostic. Again, this thread might be a good start: https://github.com/bokeh/bokeh/issues/1895

**Comment by prisae** [comment]
> And what happens if you output an object with HTML formatting in QT console? does it render it?
> 
> can you screenshot:
> 
> ```python
> import scooby
> scooby.investigate()
> ```

Missed this one. No, QtConsole does not render html, plain text output.

![Jupyter QtConsole_003](https://user-images.githubusercontent.com/8020943/60164779-76ac4700-97fe-11e9-9cbf-8a2a2d796ad2.png)

**Comment by banesullivan** [comment]
Hm... considering this is a client-side decision and there is ...

> probably no way to test client-side behaviour

... with `scooby`, then maybe we should just document that `in_jupyter` works most of the time but sometimes it's difficult to know what's under the mask

![2eeunw](https://user-images.githubusercontent.com/22067021/60165730-0df0aa80-97bd-11e9-8a35-9dc6e9a460db.jpg)

**Comment by prisae** [comment]
Maybe. But I like the `in_kernel` or `has_kernel` or `uses_kernel` more, as it is what it is, we check if there is a kernel. Then we could say that a kernel usually means jupyter notebook or jupyter lab, but it can also be other frontends which use the kernel (and might not be able to render html).

**Comment by banesullivan** [comment]
I agree. How about in `in_ipykernel`? and we'll scrap the `in_jupyter` method until there becomes a solution.

**Comment by prisae** [comment]
Other use-case: IPython-console in Jupyter-lab:

![Selection_001](https://user-images.githubusercontent.com/8020943/60166244-19fe5b80-9801-11e9-9611-84b0cc9adc36.png)

So here even in IPython it says `True`, because it uses the same kernel. A kernel can be used by any frontend...

**Comment by prisae** [comment]
And to complicate things, IPython in JupyterLab CAN render html...

![Selection_002](https://user-images.githubusercontent.com/8020943/60166345-4ca85400-9801-11e9-935a-ea35d534cc21.png)

**Comment by banesullivan** [comment]
This is becoming way more complicated than I could have ever imagined...

I'm still thinking that there might be a way to tell if HTML is rendered or not... but I'll save this for when I have free time

**Comment by prisae** [comment]
:+1: for `in_ipykernel`

**Comment by prisae** [comment]
And yet another complication: You can start IPython with an existing kernel from a notebook or console by running
```
jupyter console --existing
```
So you are in the same IPython as otherwise, but `in_jupyter` will yield `True`, yet it cannot render html:

![dtr: ~_003](https://user-images.githubusercontent.com/8020943/60166867-34850480-9802-11e9-9200-6c5668095c6a.png)

**Comment by prisae** [comment]
Believe me, I tried many things... :rofl:
